src/core/Akka.Streams/Implementation/Sinks.cs

```diff
@@ -1118,4 +1120,96 @@ public override ILogicAndMaterializedValue<Task<TMat>> CreateLogicAndMaterialize
         /// <returns>TBD</returns>
         public override string ToString() => "LazySink";
     }
+
+    internal sealed class ObservableSinkStage<T> : GraphStageWithMaterializedValue<SinkShape<T>, IObservable<T>>
+    {
+        #region internal classes
+
+        private sealed class ObserverDisposable : IDisposable
+        {
+            private readonly ObservableLogic _logic;
+            private readonly IObserver<T> _observer;
+            private readonly AtomicBoolean _disposed = new AtomicBoolean(false);
+
+            public ObserverDisposable(ObservableLogic logic, IObserver<T> observer)
+            {
+                _logic = logic;
+                _observer = observer;
+            }
+
+            public void Dispose() => Dispose(unregister: true);
+
+            public void Dispose(bool unregister)
+            {
+                if (_disposed.CompareAndSet(false, true))
+                {
+                    if (unregister) _logic.Remove(_observer);
+
+                    _observer.OnCompleted();
+                }
+                else
+                {
+                    throw new ObjectDisposedException("ObservableSink subscription has been already disposed.");
+                }
+            }
+        }
+        
+        private sealed class ObservableLogic : GraphStageLogic, IObservable<T>
+        {
+            private readonly ObservableSinkStage<T> _stage;
+            private ImmutableDictionary<IObserver<T>, ObserverDisposable> _observers = ImmutableDictionary<IObserver<T>, ObserverDisposable>.Empty;
+
+            public ObservableLogic(ObservableSinkStage<T> stage) : base(stage.Shape)
+            {
+                _stage = stage;
+                SetHandler(stage.Inlet,
+                    onPush: () =>
+                    {
+                        var element = Grab(stage.Inlet);
+                        foreach (var observer in _observers.Keys) observer.OnNext(element);
+
+                        Pull(stage.Inlet);
+                    },
+                    onUpstreamFinish: () =>
+                    {
+                        var old = Interlocked.Exchange(ref _observers, ImmutableDictionary<IObserver<T>, ObserverDisposable>.Empty);
+                        foreach (var disposer in old.Values) disposer.Dispose(unregister: false);
+                    },
+                    onUpstreamFailure: e =>
+                    {
```

Comment:
Don’t we need to remove observers in case of failure too?

Reply:
Are we 100% sure that upstream failure should cause removal? What about `Rescue`?

Reply:
Even if the observable catches the failure the Sink is dead so it wouldn’t produce any new items